toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -48,6 +52,164 @@ auto VariantMatch(V&& v, Fs&&... fs) -> decltype(auto) {
+  // fall back to a UTF-8 decoding path (not yet fully implemented).
+  //
+  // bit6 remains unused if we need to classify more characters.
+  const auto high_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b0000'0000,
+      /*0x1:*/ 0b0000'0000,
+      /*0x2:*/ 0b0000'0000,
+      /*0x3:*/ 0b0000'0010,
+      /*0x4:*/ 0b0000'0100,
+      /*0x5:*/ 0b0000'1001,
+      /*0x6:*/ 0b0001'0000,
+      /*0x7:*/ 0b0010'0000,
+      /*0x8:*/ 0b1000'0000,
+      /*0x9:*/ 0b1000'0000,
+      /*0xA:*/ 0b1000'0000,
+      /*0xB:*/ 0b1000'0000,
+      /*0xC:*/ 0b1000'0000,
+      /*0xD:*/ 0b1000'0000,
+      /*0xE:*/ 0b1000'0000,
+      /*0xF:*/ 0b1000'0000);
+  const auto low_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b1010'1010,
+      /*0x1:*/ 0b1011'1110,
+      /*0x2:*/ 0b1011'1110,
+      /*0x3:*/ 0b1011'1110,
+      /*0x4:*/ 0b1011'1110,
+      /*0x5:*/ 0b1011'1110,
+      /*0x6:*/ 0b1011'1110,
+      /*0x7:*/ 0b1011'1110,
+      /*0x8:*/ 0b1011'1110,
+      /*0x9:*/ 0b1011'1110,
+      /*0xA:*/ 0b1011'1100,
+      /*0xB:*/ 0b1001'0100,
+      /*0xC:*/ 0b1001'0100,
+      /*0xD:*/ 0b1001'0100,
+      /*0xE:*/ 0b1001'0100,
+      /*0xF:*/ 0b1001'0101);
+
+  int tail_utf8_mask;
+  int tail_ascii_mask;
+  int tail_mask;
+  // Use `ssize_t` for performance here as we index memory in a tight loop.
+  ssize_t i = 0;
+  const ssize_t size = text.size();
+  while ((i + 16) <= size) {
+    __m128i input =
+        _mm_loadu_si128(reinterpret_cast<const __m128i*>(text.data() + i));
+    __m128i low_mask = _mm_shuffle_epi8(low_lut, input);
+    __m128i input_high =
+        _mm_and_si128(_mm_srli_epi32(input, 4), _mm_set1_epi8(0x0f));
+    __m128i high_mask = _mm_shuffle_epi8(high_lut, input_high);
+    __m128i mask = _mm_and_si128(low_mask, high_mask);
+
+    // The high bits of the mask directly mark any non-ASCII UTF-8 code units,
+    // extract those first.
+    // TODO: This is technically unnecessary as we ignore Unicode characters
+    // below, but we should eventually support them and this keeps the core
+    // optimized loop set-up to handle them.
+    tail_utf8_mask = _mm_movemask_epi8(mask);
+
+    // Now compare to find the completely zero bytes.
+    __m128i id_byte_mask_vec = _mm_cmpeq_epi8(mask, _mm_setzero_si128());
+    tail_ascii_mask = _mm_movemask_epi8(id_byte_mask_vec);
+
+    // We use the OR of the UTF-8 mask and the ASCII mask to find the known ID
+    // range.
+    tail_mask = tail_ascii_mask | tail_utf8_mask;
+
+    // Check if there are bits in the tail mask, which means zero bytes and the
+    // end of the identifier. We could do this without materializing the scalar
+    // mask on more recent CPUs, but we generally expect the median length we
+    // encounter to be <16 characters and so we avoid the extra instruction in
+    // that case and predict this branch to succeed so it is laid out in a
+    // reasonable way.
+    if (LLVM_LIKELY(tail_mask != 0)) {
+      break;
+    }
+    i += 16;
+  }
+  if (LLVM_LIKELY((i + 16) <= size)) {
```

Comment:
Rather than repeating the branch condition, I think it might be a bit clearer to move the `__builtin_ctz` above the `break` a few lines above.

Reply:
Sure, good idea. and with a much simpler duplication, the fallback path is much cleaner I think, and we now have a single place to introduce unicode handling eventually.